In [120]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score

In [121]:
df=pd.read_excel(r"D:\Motor Trend Car Road Tests (1).xlsx")

In [122]:
df.describe()

mpg        cyl        disp          hp       drat         wt  \
count  32.000000  32.000000   32.000000   32.000000  32.000000  32.000000   
mean   20.090625   6.187500  230.721875  146.687500   3.596563   3.217250   
std     6.026948   1.785922  123.938694   68.562868   0.534679   0.978457   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.513000   
25%    15.425000   4.000000  120.825000   96.500000   3.080000   2.581250   
50%    19.200000   6.000000  196.300000  123.000000   3.695000   3.325000   
75%    22.800000   8.000000  326.000000  180.000000   3.920000   3.610000   
max    33.900000   8.000000  472.000000  335.000000   4.930000   5.424000   

            qsec         vs         am       gear     carb  
count  32.000000  32.000000  32.000000  32.000000  32.0000  
mean   17.848750   0.437500   0.406250   3.687500   2.8125  
std     1.786943   0.504016   0.498991   0.737804   1.6152  
min    14.500000   0.000000   0.000000   3.000000   1.0000  
25%    16.892500   0.000000   0.000000   3.000000   2.0000  
50%    17.710000   0.000000   0.000000   4.000000   2.0000  
75%    18.900000   1.000000   1.000000   4.000000   4.0000  
max    22.900000   1.000000   1.000000   5.000000   8.0000

**1.1**

In [123]:
df2=df.drop(columns='model')
df2.columns


Index(['mpg', 'cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear',
       'carb'],
      dtype='object')

In [124]:
cols=df2.columns
cols=cols.drop('mpg')

x_list = []

for col in cols:
    x_list.append(df2[col].values.reshape(-1,1))

n=len(df2)                                      
ones=np.ones([n,1])
x=np.hstack(x_list)
x=np.hstack([ones,x])
y=df2['mpg']
modelo=sm.OLS(y,x)
results=modelo.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        15:42:30   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.3034     18.718      0.657      0.518     -26.623      51.229
x1            -0.1114      1.045     -0.107      0.916      -2.285       2.062
x2             0.0133      0.018      0.747      0.463      -0.024       0.050
x3            -0.0215      0.022     -0.987      0.335      -0.067       0.024
x4             0.7871      1.635      0.481      0.635      -2.614       4.188
x5            -3.7153      1.894     -1.961      0.063      -7.655       0.224
x6             0.8210      0.731      1.123      0.274      -0.699       2.341
x7             0.3178      2.105      0.151      0.881      -4.059       4.694
x8             2.5202      2.057      1.225      0.234      -1.757       6.797
x9             0.6554      1.493      0.439      0.665      -2.450       3.761
x10           -0.1994      0.829     -0.241      0.812      -1.923       1.524
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                     1.22e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.22e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [125]:
xtrain,xtest,ytrain,ytest=train_test_split(x,y,train_size=0.40,random_state=137)
scaler=StandardScaler()
Xscaled_train=scaler.fit_transform(xtrain)
ones=np.ones((len(xtrain),1))
X_e_train=np.hstack((ones,Xscaled_train))
modelo=sm.OLS(ytrain,X_e_train)
results=modelo.fit()
results.summary()

c:\Users\marce\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Wed, 17 Sep 2025   Prob (F-statistic):              0.242
Time:                        15:42:30   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.020      10.370      24.497
x1         -3.084e-15    3.4e-15     -0.906      0.531   -4.63e-14    4.02e-14
x2            40.7571     54.010      0.755      0.588    -645.499     727.013
x3            -2.4635      3.021     -0.815      0.565     -40.854      35.927
x4           -10.1581     19.484     -0.521      0.694    -257.727     237.411
x5            14.1135     13.469      1.048      0.485    -157.022     185.249
x6            -0.8005      3.106     -0.258      0.839     -40.266      38.665
x7             1.5727      3.475      0.453      0.729     -42.583      45.729
x8            32.4926     41.262      0.787      0.575    -491.795     556.780
x9            36.1394     42.669      0.847      0.553    -506.022     578.301
x10          -38.3551     43.625     -0.879      0.541    -592.668     515.958
x11           11.9707     17.482      0.685      0.618    -210.157     234.098
==============================================================================
Omnibus:                        0.259   Durbin-Watson:                   2.354
Prob(Omnibus):                  0.879   Jarque-Bera (JB):                0.413
Skew:                           0.212   Prob(JB):                        0.813
Kurtosis:                       2.196   Cond. No.                     4.39e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.05e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Aca los coeficientes de son enormes, los mas grandes son x2, x9 y x10

In [126]:
Xscaled_test = scaler.transform(xtest)
x_test_scaled = sm.add_constant(Xscaled_test)
y_pred_test = results.predict(x_test_scaled)
r2_test = r2_score(ytest, y_pred_test)
print(f"R^2 Train: 0.990")
print(f"R^2 Test: {r2_test}")

R^2 Train: 0.990
R^2 Test: -54.0717852041876


In [127]:
from sklearn.linear_model import Ridge


In [128]:
ridge=Ridge(alpha=1)
ridge.fit(x,y)
print(f"Coeficientes:{ ridge.coef_}Intercepto:{ridge.intercept_} R^2:{ridge.score(x,y)}")

Coeficientes:[ 0.         -0.29040511  0.00418935 -0.01788854  0.82149554 -2.62359314
  0.44301974  0.19726344  1.75515664  0.82884165 -0.51174214]Intercepto:18.70739008719543 R^2:0.8654505729609845


**1.2**

In [129]:
cols=df2.columns
cols=cols.drop('qsec')

x_list = []

for col in cols:
    x_list.append(df2[col].values.reshape(-1,1))

n=len(df2)                                      
ones=np.ones([n,1])
x=np.hstack(x_list)
x=np.hstack([ones,x])
y=df2['qsec']
modelo=sm.OLS(y,x)
results=modelo.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        15:42:30   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.7762      3.876      4.586      0.000       9.716      25.837
x1             0.0690      0.061      1.123      0.274      -0.059       0.197
x2            -0.3627      0.293     -1.239      0.229      -0.971       0.246
x3            -0.0075      0.005     -1.505      0.147      -0.018       0.003
x4            -0.0016      0.006     -0.242      0.811      -0.015       0.012
x5            -0.1311      0.476     -0.275      0.786      -1.121       0.859
x6             1.4963      0.500      2.990      0.007       0.456       2.537
x7             0.9700      0.573      1.694      0.105      -0.221       2.161
x8            -0.9012      0.585     -1.540      0.139      -2.118       0.316
x9            -0.2013      0.433     -0.465      0.647      -1.101       0.699
x10           -0.2736      0.233     -1.174      0.254      -0.758       0.211
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                     8.77e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.77e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [130]:
xtrain,xtest,ytrain,ytest=train_test_split(x,y,train_size=0.40,random_state=137)
scaler=StandardScaler()
Xscaled_train=scaler.fit_transform(xtrain)
ones=np.ones((len(xtrain),1))
X_e_train=np.hstack((ones,Xscaled_train))
modelo=sm.OLS(ytrain,X_e_train)
results=modelo.fit()
results.summary()

c:\Users\marce\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     4.609
Date:                Wed, 17 Sep 2025   Prob (F-statistic):              0.349
Time:                        15:42:30   Log-Likelihood:               -0.25847
No. Observations:                  12   AIC:                             22.52
Df Residuals:                       1   BIC:                             27.85
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733      0.247     70.269      0.009      14.232      20.515
x1          2.057e-15   1.14e-13      0.018      0.988   -1.44e-12    1.45e-12
x2             1.0235      2.262      0.453      0.729     -27.713      29.760
x3            -0.3081     30.092     -0.010      0.993    -382.663     382.047
x4             0.2789      1.711      0.163      0.897     -21.465      22.023
x5            -1.0127      9.720     -0.104      0.934    -124.519     122.494
x6            -1.0356      8.615     -0.120      0.924    -110.497     108.426
x7             0.5976      1.295      0.461      0.725     -15.861      17.057
x8             0.2129     23.358      0.009      0.994    -296.578     297.004
x9            -0.3877     24.867     -0.016      0.990    -316.349     315.574
x10            0.6755     25.827      0.026      0.983    -327.485     328.836
x11           -0.1562      9.422     -0.017      0.989    -119.876     119.563
==============================================================================
Omnibus:                       12.359   Durbin-Watson:                   2.150
Prob(Omnibus):                  0.002   Jarque-Bera (JB):                7.063
Skew:                           1.532   Prob(JB):                       0.0293
Kurtosis:                       5.178   Cond. No.                     1.60e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.57e-33. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [131]:
Xscaled_test = scaler.transform(xtest)
x_test_scaled = sm.add_constant(Xscaled_test)
y_pred_test = results.predict(x_test_scaled)
r2_test = r2_score(ytest, y_pred_test)
print(f"R^2 Test: {r2_test}")

R^2 Test: 0.37610389620006013


In [132]:
ridge=Ridge(alpha=0.5)
ridge.fit(x,y)
print(f"Coeficientes:{ ridge.coef_}Intercepto:{ridge.intercept_} R^2:{ridge.score(x,y)}")

Coeficientes:[ 0.          0.05622932 -0.40955423 -0.00616575 -0.0028698  -0.16850012
  1.31164044  0.88966376 -0.83041426 -0.26595415 -0.23079392]Intercepto:19.060808652249285 R^2:0.8729730722946962


**2.1**

In [133]:
dfd=pd.get_dummies(df2,columns=['cyl','gear','carb'])

dfd.head(32)

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0   21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1   21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2   22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3   21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4   18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   
5   18.1  225.0  105  2.76  3.460  20.22   1   0  False   True  False    True   
6   14.3  360.0  245  3.21  3.570  15.84   0   0  False  False   True    True   
7   24.4  146.7   62  3.69  3.190  20.00   1   0   True  False  False   False   
8   22.8  140.8   95  3.92  3.150  22.90   1   0   True  False  False   False   
9   19.2  167.6  123  3.92  3.440  18.30   1   0  False   True  False   False   
10  17.8  167.6  123  3.92  3.440  18.90   1   0  False   True  False   False   
11  16.4  275.8  180  3.07  4.070  17.40   0   0  False  False   True    True   
12  17.3  275.8  180  3.07  3.730  17.60   0   0  False  False   True    True   
13  15.2  275.8  180  3.07  3.780  18.00   0   0  False  False   True    True   
14  10.4  472.0  205  2.93  5.250  17.98   0   0  False  False   True    True   
15  10.4  460.0  215  3.00  5.424  17.82   0   0  False  False   True    True   
16  14.7  440.0  230  3.23  5.345  17.42   0   0  False  False   True    True   
17  32.4   78.7   66  4.08  2.200  19.47   1   1   True  False  False   False   
18  30.4   75.7   52  4.93  1.615  18.52   1   1   True  False  False   False   
19  33.9   71.1   65  4.22  1.835  19.90   1   1   True  False  False   False   
20  21.5  120.1   97  3.70  2.465  20.01   1   0   True  False  False    True   
21  15.5  318.0  150  2.76  3.520  16.87   0   0  False  False   True    True   
22  15.2  304.0  150  3.15  3.435  17.30   0   0  False  False   True    True   
23  13.3  350.0  245  3.73  3.840  15.41   0   0  False  False   True    True   
24  19.2  400.0  175  3.08  3.845  17.05   0   0  False  False   True    True   
25  27.3   79.0   66  4.08  1.935  18.90   1   1   True  False  False   False   
26  26.0  120.3   91  4.43  2.140  16.70   0   1   True  False  False   False   
27  30.4   95.1  113  3.77  1.513  16.90   1   1   True  False  False   False   
28  15.8  351.0  264  4.22  3.170  14.50   0   1  False  False   True   False   
29  19.7  145.0  175  3.62  2.770  15.50   0   1  False   True  False   False   
30  15.0  301.0  335  3.54  3.570  14.60   0   1  False  False   True   False   
31  21.4  121.0  109  4.11  2.780  18.60   1   1   True  False  False   False   

    gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0     True   False   False   False   False    True   False   False  
1     True   False   False   False   False    True   False   False  
2     True   False    True   False   False   False   False   False  
3    False   False    True   False   False   False   False   False  
4    False   False   False    True   False   False   False   False  
5    False   False    True   False   False   False   False   False  
6    False   False   False   False   False    True   False   False  
7     True   False   False    True   False   False   False   False  
8     True   False   False    True   False   False   False   False  
9     True   False   False   False   False    True   False   False  
10    True   False   False   False   False    True   False   False  
11   False   False   False   False    True   False   False   False  
12   False   False   False   False    True   False   False   False  
13   False   False   False   False    True   False   False   False  
14   False   False   False   False   False    True   False   False  
15   False   False   False   False   False    True   False   False  
16   False   False   False   False   False    True   False   False  
17    True   False    True   False   False   False   False   False  
18    True   False  

In [134]:
cols=dfd.columns
cols=cols.drop('mpg')
x_list = []

for col in cols:
    x_list.append(dfd[col].values.reshape(-1,1))

n=len(dfd)                                      
ones=np.ones([n,1])
x=np.hstack(x_list)
x=np.hstack([ones,x])
x = np.asarray(x, dtype=float)  
y=dfd['mpg']
modelo=sm.OLS(y,x)
results=modelo.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           0.000124
Time:                        15:42:30   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.4937     10.672      1.358      0.195      -8.254      37.241
x1             0.0355      0.032      1.114      0.283      -0.032       0.104
x2            -0.0705      0.039     -1.788      0.094      -0.155       0.014
x3             1.1828      2.483      0.476      0.641      -4.111       6.476
x4            -4.5298      2.539     -1.784      0.095      -9.941       0.881
x5             0.3678      0.935      0.393      0.700      -1.626       2.362
x6             1.9309      2.871      0.672      0.512      -4.189       8.051
x7             1.2121      3.214      0.377      0.711      -5.637       8.062
x8             5.8262      5.651      1.031      0.319      -6.218      17.871
x9             3.1775      3.526      0.901      0.382      -4.337      10.692
x10            5.4900      4.884      1.124      0.279      -4.920      15.900
x11            3.6170      3.548      1.020      0.324      -3.945      11.179
x12            4.7313      4.622      1.024      0.322      -5.120      14.583
x13            6.1454      4.099      1.499      0.155      -2.591      14.882
x14           -0.0577      3.844     -0.015      0.988      -8.250       8.135
x15           -1.0370      2.923     -0.355      0.728      -7.267       5.193
x16            2.9420      3.393      0.867      0.400      -4.291      10.175
x17            1.0338      2.628      0.393      0.700      -4.567       6.634
x18            4.4199      4.041      1.094      0.291      -4.193      13.033
x19            7.1927      5.924      1.214      0.243      -5.434      19.820
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     1.21e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.04e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

la variable mas importante es la ultima pero tenemos una constante enorme 

In [135]:
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=0.40, random_state=137)
scaler = StandardScaler().fit(xtrain)
Xscaled_train = scaler.transform(xtrain)
Xscaled_test = scaler.transform(xtest)
X_e_train = sm.add_constant(Xscaled_train) 
X_e_test = sm.add_constant(Xscaled_test)
modelo = sm.OLS(ytrain, X_e_train)
results = modelo.fit()
y_pred_train = results.predict(X_e_train)
r2_train = r2_score(ytrain, y_pred_train)
r2_test = r2_score(ytest, y_pred_test)
print(f"R² Train: {r2_train}")
print(f"R² Test: {r2_test}")


R² Train: 1.0
R² Test: -0.13944952797158994


no hay mucho que decir, el overfit esta brutal

**2.2**

In [136]:
cols=dfd.columns
cols=cols.drop('qsec')
x_list = []

for col in cols:
    x_list.append(dfd[col].values.reshape(-1,1))

n=len(dfd)                                      
ones=np.ones([n,1])
x=np.hstack(x_list)
x=np.hstack([ones,x])
x = np.asarray(x, dtype=float)  
y=dfd['qsec']
modelo=sm.OLS(y,x)
results=modelo.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Wed, 17 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        15:42:30   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.9157      2.339      3.385      0.004       2.931      12.900
x1             0.0277      0.071      0.393      0.700      -0.123       0.178
x2             0.0035      0.009      0.389      0.703      -0.016       0.023
x3            -0.0021      0.012     -0.173      0.865      -0.027       0.023
x4             0.1079      0.687      0.157      0.877      -1.356       1.571
x5             0.8105      0.739      1.097      0.290      -0.764       2.385
x6             0.2657      0.797      0.333      0.744      -1.434       1.965
x7            -1.6943      0.771     -2.197      0.044      -3.338      -0.050
x8             3.9957      1.231      3.247      0.005       1.373       6.619
x9             2.8913      0.656      4.405      0.001       1.492       4.290
x10            1.0288      1.371      0.750      0.465      -1.894       3.951
x11            2.1337      0.843      2.530      0.023       0.336       3.931
x12            3.4660      0.961      3.608      0.003       1.418       5.514
x13            2.3160      1.049      2.209      0.043       0.081       4.551
x14            2.3040      0.872      2.642      0.018       0.445       4.163
x15            1.4696      0.711      2.066      0.057      -0.046       2.985
x16            2.0747      0.791      2.624      0.019       0.390       3.760
x17            0.3570      0.719      0.496      0.627      -1.176       1.890
x18            0.7388      1.137      0.650      0.526      -1.685       3.163
x19            0.9717      1.686      0.576      0.573      -2.623       4.566
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     7.46e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 5.33e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [137]:
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=0.40, random_state=137)
scaler = StandardScaler().fit(xtrain)
Xscaled_train = scaler.transform(xtrain)
Xscaled_test = scaler.transform(xtest)
X_e_train = sm.add_constant(Xscaled_train) 
X_e_test = sm.add_constant(Xscaled_test)
modelo = sm.OLS(ytrain, X_e_train)
results = modelo.fit()
y_pred_train = results.predict(X_e_train)
r2_train = r2_score(ytrain, y_pred_train)
r2_test = r2_score(ytest, y_pred_test)
print(f"R² Train: {r2_train}")
print(f"R² Test: {r2_test}")

R² Train: 1.0
R² Test: 0.37610389620006013


es lo mismo, es un overfit enorme, me parece increible que da 1 en la $R^2$ del overfitting que se hace

**3.1**

hay mas cesgo en el ejercicio de 1.1, al convertir la data cualitativa a cuantitativa se vuelve un muchisimo mejor modelo pero el que nuestra n sea 32 y que eligamos solo el 40% de train hace que no haya mucho que entrenar, al hacer el get_dummies llega a haber mas variables predictoras que datos.

**3.2**

sale muy parecido con la $R^2$ de ambos casos que es con y sin dummies cuando la salida es qsec, deberia de a ver una mejor correlacion al quitar las variables cualitativas pero es muy parecido el resultado, tambien hay un overfit muy fuerte muy alto.